### DISCUSSION TITLE
Question: How to do an interactive shell input like bash?

### DISCUSSION THREAD

**Discussion by ZephyrPhrost** [charmbracelet/bubbletea#1173] [Q&A]
**Describe the bug**

I'm trying to do an interactive shell like it's was possible with: https://github.com/c-bata/go-prompt

but I'm struggling a bit with the view model:

```go
// Update handles incoming messages.
func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var cmd tea.Cmd

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.Type {
		case tea.KeyEnter:
			input := strings.TrimSpace(m.input.Value())
			if input != "" {
				if input == "exit" {
					return m, tea.Quit
				}
				response := m.bot.ExecuteCommand(input)
				m.output = response
				m.input.SetValue("")
			} else {
				m.output += "\n"
			}
		case tea.KeyCtrlC:
			return m, tea.Quit
		}
	}

	// Update the text input component
	m.input, cmd = m.input.Update(msg)
	return m, cmd
}

// View renders the UI.
func (m model) View() string {
	var b strings.Builder

	// Output Section
	b.WriteString(m.styles.outputStyle.Render(m.output))

	// Input Section
	b.WriteString("\n")
	b.WriteString(m.input.View()) // No extra "> " prefix

	return b.String()
}
```

I wonder if there is an example with an interactive shell?

I'm expecting the following when I press enter:

```
>
>
>
```

Similar to what's possible in: https://github.com/c-bata/go-prompt/tree/master/_example

**Comment by odino** [comment]
* store inputs and outputs in a `messages []string{}` on your model
* on the view, `strings.Join(m.messages, "\n")`

🙏

**Reply by odino** [reply]
BTW the [suggestion](https://github.com/charmbracelet/bubbletea/discussions/1173#discussioncomment-12785221) from @meowgorithm is what you should do -- I wasnt aware of Println (and obviously trying to repaint n []messages{} will be painful).

You can see kind of what you need here: https://github.com/abs-lang/abs/blob/new-terminal/terminal/terminal.go

(I'm building a repl that evaluates some code so there's a lot of overlap with your original use case)

**Reply by odino** [reply]
FYI you can see an actual REPL here: https://github.com/abs-lang/abs/releases/tag/2.7.0
I also used go-prompt earlier on, and I must admit the switch to bubbletea has been a blessing 🙏

**Comment by meowgorithm** [comment]
You're going to want to print output above the interactive portion of the program managed by Bubble Tea so it flows into the terminal's scroll back butter. For that have a look at the [`tea.Println`](https://pkg.go.dev/github.com/charmbracelet/bubbletea#Println) `Cmd` as well the [package manger](https://github.com/charmbracelet/bubbletea/tree/main/examples/package-manager) example.

Basically:

```go
switch msg {
case somethingWasExecuted:
	return m, tea.Println("> some output")
}
```
